### PR TITLE
For #23876 - Update kotlinx-coroutines to 1.6.1. Refactor runBlockig in tests

### PR DESCRIPTION
**Pull request by Mugurell**
After the update `TestCoroutineDispatcher`, `TestCoroutineScope` and
`runBlockingTest` are deprecated.
Instead of these new `TestDispatcher`, `TestScope` and `runTest` APIs are used.

Took the oportunity to also replace most of the `runBlocking` calls with the
new `runTest` specially designed for running coroutines in tests API.

Needs https://github.com/mozilla-mobile/android-components/pull/12006
All tests pass locally.

----------

Copying here my reasoning described on AC:
- goal was to refactor out `TestCoroutineScope`, `TestCoroutineDispatcher` and `runBlockingTests` since they are deprecated and would fail the build (if not suppressing the deprecation)
- as recommended I also tried to remove as many `runBlocking` usages as possible but in some cases with more intricate scenarios where interacting with other dispatchers / threads was needed some `runBlocking` calls were kept.
- `runTest` replaces all `runBlocking` and `runBlockingTest` usages.
`TestScope`, `UnconfinedTestDispatcher` and `StandardTestDispatcher` replaces all usages of `TestCoroutineScope`, `TestCoroutineDispatcher` as per the official documentation.
- `runTestOnMain` was used where we'd set the main dispatcher to be able to syncronize runs/code in tests/system under test. Occasionally when the `runTest` replacement would have the test fail I added the `MainCoroutineTest` rule and saw that the test passed with also using `runTestOnMain`.
- there are no modifications in the code being tested and no modifications in the tests logic. This seems like a good guarantee that if all tests pass on CI then the changes are good.

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

File: app/src/test/java/org/mozilla/fenix/exceptions/login/LoginExceptionsInteractorTest.kt
Comment (MatthewTighe) [inline app/src/test/java/org/mozilla/fenix/exceptions/login/LoginExceptionsInteractorTest.kt:30]: Just out of curiosity, why are `UnconfinedTestDispatcher` and the extension function version of `runTest` required for this test class?

File: app/src/test/java/org/mozilla/fenix/gleanplumb/OnDiskMessageMetadataStorageTest.kt
Comment (MatthewTighe) [inline app/src/test/java/org/mozilla/fenix/gleanplumb/OnDiskMessageMetadataStorageTest.kt:47]: Also curious about the usage of `UnconfinedTestDispatcher` here.

**Review by MatthewTighe — COMMENTED** [review]
Had a couple questions/comments. LGTM otherwise, I'll come back for an approving review once CI is passing after [12006](https://github.com/mozilla-mobile/android-components/pull/12006).

File: app/src/test/java/org/mozilla/fenix/exceptions/login/LoginExceptionsInteractorTest.kt
Comment (Mugurell) [reply app/src/test/java/org/mozilla/fenix/exceptions/login/LoginExceptionsInteractorTest.kt:30]: There are two dispatchers to choose from: `UnconfinedTestDispatcher` and `StandardTestDispatcher`.
`UnconfinedTestDispatcher` has the nice advantage (here) that it eagerly enters `launch` and `async` blocks.
We need this since the interactor executes for example
```
ioScope.launch {
    loginExceptionStorage.removeLoginException(item)
}
```
Using the same scope (and dispatcher) in the interactor and for running the test allows to sync the execution.

File: app/src/test/java/org/mozilla/fenix/gleanplumb/OnDiskMessageMetadataStorageTest.kt
Comment (Mugurell) [reply app/src/test/java/org/mozilla/fenix/gleanplumb/OnDiskMessageMetadataStorageTest.kt:47]: Tested now. Here it doesn't seem to be needed. Doesn't break anything but is not required.
Will remove it in favor of the default `StandardTestDispatcher`.

File: app/src/test/java/org/mozilla/fenix/settings/deletebrowsingdata/DefaultDeleteBrowsingDataControllerTest.kt
Comment (Mugurell) [reply app/src/test/java/org/mozilla/fenix/settings/deletebrowsingdata/DefaultDeleteBrowsingDataControllerTest.kt:60]: Thanks. This was probably just a quick `use runTestOnMain anywhere MainCoroutineRule is set` though I missed setting the appropriate context.
Just to enforce this automatism I think the best way to reconcile this is to replace the context and keep runTestOnMain.

**Comment by Mugurell** [comment]
Rebased to fix conflicts, replaced a few more `runBlocking` occurences.

File: app/src/test/java/org/mozilla/fenix/exceptions/login/LoginExceptionsInteractorTest.kt
Comment (csadilek) [reply app/src/test/java/org/mozilla/fenix/exceptions/login/LoginExceptionsInteractorTest.kt:30]: Yes, switching to `UnconfinedTestDispatcher` does make the nested coroutines launch, but they can launch on a different thread and we're not synchronizing the execution i.e., the test code is not waiting for their completion. For many tests this will not be good enough and lead to intermittent failures. I've added and example here: https://github.com/mozilla-mobile/android-components/pull/12006#pullrequestreview-960776443

It's easy to fix, but we should review the tests that launch nested coroutines and make assertion based on them.

**Comment by Mugurell** [comment]
CI is green. Rebasing and trying again.

**Review by csadilek — APPROVED** [review]
We've worked through this in A-C and this follows the same pattern: https://github.com/mozilla-mobile/android-components/pull/12006#pullrequestreview-966690533

🚢

**Comment by csadilek** [comment]
💚

**Pull request merged**
